core/src/toga/app.py

```diff
@@ -192,137 +243,112 @@ def __init__(
 
-        # Keep an accessible copy of the app instance
+        # Keep an accessible copy of the app singleton instance
         App.app = self
 
         # We need a module name to load app metadata. If an app_name has been
         # provided, we can set the app name now, and derive the module name
         # from there.
+        self._app_name = app_name
         if app_name:
-            self._app_name = app_name
+            metadata_module_name = self.module_name
         else:
-            # If the code is contained in appname.py, and you start the app
-            # using `python -m appname`, the main module package will report
-            # as ''. Set the initial app name as None.
+            # If the code is contained in appname.py, and you start the app using
+            # `python -m appname`, the main module package will report as ''. Set the
+            # metadata module name as None.
             #
-            # If the code is contained in appname.py, and you start the app
-            # using `python appname.py`, the main module will report as None.
+            # If the code is contained in appname.py, and you start the app using
+            # `python appname.py`, the metadata module name will report as None.
             #
-            # If the code is contained in a folder, and you start the app
-            # using `python -m appname`, the main module will report as the
-            # name of the folder.
+            # If the code is contained in a folder, and you start the app using `python
+            # -m appname`, the metadata module name will report as the name of the
+            # folder.
             try:
                 main_module_pkg = sys.modules["__main__"].__package__
                 if main_module_pkg == "":
-                    self._app_name = None
+                    metadata_module_name = None
                 else:
-                    self._app_name = main_module_pkg
+                    metadata_module_name = main_module_pkg
             except KeyError:
                 # We use the existence of a __main__ module as a proxy for
                 # being in test conditions. This isn't *great*, but the __main__
                 # module isn't meaningful during tests, and removing it allows
                 # us to avoid having explicit "if under test conditions" checks.
                 # If there's no __main__ module, we're in a test, and we can't
                 # imply an app name from that module name.
-                self._app_name = None
+                metadata_module_name = None
 
-            # Try deconstructing the app name from the app ID
-            if self._app_name is None and app_id:
-                self._app_name = app_id.split(".")[-1]
+            # Try deconstructing the metadata module name from the app ID
+            if metadata_module_name is None and app_id:
+                metadata_module_name = app_id.split(".")[-1].replace("-", "_")
```

Comment:
If I'm reading this right, we don't need to do this denormalisation step - we can just use the last name part.

Reply:
Done.